File: shared/notify/message.go
package notify

type MsgType string

const (
	MsgTypeEmail MsgType = "email"
	MsgTypSMS    MsgType = "sms"
)

type Message interface {
	message()
}

type TemplateMessage struct {
	Type     MsgType        `json:"type"`
	To       []string       `json:"to"`
	Content  map[string]any `json:"content"`
	Template string         `json:"template"`
}

func (*TemplateMessage) message() {}

type SimpleMessage struct {
	Type    MsgType  `json:"type"`
	To      []string `json:"to"`
	Content string   `json:"content"`
}

func (*SimpleMessage) message() {}

File: notification-service/src/dto/message.ts
import { z } from "zod";

export const MessageSchema = z.object({
    type: z.enum(["email", "sms"]),
    to: z.array(z.string()).min(1),
    template: z.string().optional(),
    content: z.union([z.string(), z.any()])
}).refine((o) => typeof o.content === "string" || !!o.template, { message: "Invalid message type" });



export type Message = z.infer<typeof MessageSchema>;